File: app/src/main/java/com/o4x/musical/ui/adapter/GenreAdapter.java
package com.o4x.musical.ui.adapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.o4x.musical.R;
import com.o4x.musical.helper.SortOrder;
import com.o4x.musical.helper.menu.SongsMenuHelper;
import com.o4x.musical.imageloader.universalil.listener.PaletteImageLoadingListener;
import com.o4x.musical.imageloader.universalil.loader.UniversalIL;
import com.o4x.musical.interfaces.CabHolder;
import com.o4x.musical.model.Genre;
import com.o4x.musical.model.Song;
import com.o4x.musical.ui.adapter.artist.ArtistAdapter;
import com.o4x.musical.ui.adapter.base.AbsAdapter;
import com.o4x.musical.ui.adapter.base.AbsMultiSelectAdapter;
import com.o4x.musical.ui.adapter.base.MediaEntryViewHolder;
import com.o4x.musical.util.MusicUtil;
import com.o4x.musical.util.NavigationUtil;
import com.o4x.musical.util.PreferenceUtil;
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import code.name.monkey.appthemehelper.util.ColorUtil;
import code.name.monkey.appthemehelper.util.MaterialValueHelper;

public class GenreAdapter extends AbsAdapter<GenreAdapter.ViewHolder, Genre> {

    public GenreAdapter(@NonNull AppCompatActivity activity, List<Genre> dataSet, @LayoutRes int itemLayoutRes, @Nullable CabHolder cabHolder) {
        super(activity, dataSet, itemLayoutRes, cabHolder);
    }


    @Override
    public long getItemId(int position) {
        return dataSet.get(position).hashCode();
    }

    @Override
    protected ViewHolder createViewHolder(@NotNull View view, int viewType) {
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);

        final Genre genre = dataSet.get(position);

        if (holder.title != null) {
            holder.title.setText(genre.getName());
        }
        if (holder.text != null) {
            holder.text.setText(MusicUtil.getGenreInfoString(activity, genre));
        }
    }

    @Override
    protected void loadImage(Genre genre, final ViewHolder holder) {
        if (holder.image == null) return;
        getImageLoader(holder)
                .load(genre)
                .into(holder.image);
    }

    @Override
    protected String getName(Genre genre) {
        return genre.getName();
    }

    @Override
    @NonNull
    protected List<Song> getSongList(@NonNull List<Genre> genres) {
        final List<Song> songs = new ArrayList<>();
        for (Genre genre : genres) {
            songs.addAll(genre.getSongs()); // maybe async in future?
        }
        return songs;
    }

    @NonNull
    @Override
    public String getSectionName(int position) {
        @Nullable String sectionName = null;
        switch (PreferenceUtil.getGenreSortOrder()) {
            case SortOrder.GenreSortOrder.GENRE_A_Z:
            case SortOrder.GenreSortOrder.GENRE_Z_A:
                sectionName = dataSet.get(position).getName();
                break;
        }

        return MusicUtil.getSectionName(sectionName);
    }

    public class ViewHolder extends MediaEntryViewHolder {

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            if (menu != null) {
                menu.setVisibility(View.GONE);
            }
        }

        @Override
        public void onClick(View view) {
            if (isInQuickSelectMode()) {
                toggleChecked(getAdapterPosition());
            } else {
                Genre genre = dataSet.get(getAdapterPosition());
                NavigationUtil.goToGenre(activity, genre);
            }
        }

        @Override
        public boolean onLongClick(View view) {
            toggleChecked(getAdapterPosition());
            return true;
        }
    }
}

File: app/src/main/java/com/o4x/musical/ui/adapter/GenreAdapter.kt
package com.o4x.musical.ui.adapter

import android.view.View
import androidx.annotation.LayoutRes
import com.o4x.musical.extensions.toGenreDetail
import com.o4x.musical.helper.SortOrder
import com.o4x.musical.interfaces.CabHolder
import com.o4x.musical.model.Genre
import com.o4x.musical.model.Song
import com.o4x.musical.ui.activities.MainActivity
import com.o4x.musical.ui.adapter.base.AbsAdapter
import com.o4x.musical.ui.adapter.base.MediaEntryViewHolder
import com.o4x.musical.util.MusicUtil
import com.o4x.musical.util.PreferenceUtil.genreSortOrder
import java.util.*

class GenreAdapter(
    val mainActivity: MainActivity,
    dataSet: List<Genre?>?,
    @LayoutRes itemLayoutRes: Int,
    cabHolder: CabHolder?
) : AbsAdapter<GenreAdapter.ViewHolder?, Genre?>(mainActivity, dataSet, itemLayoutRes, cabHolder) {

    override fun getItemId(position: Int): Long {
        return dataSet[position].hashCode().toLong()
    }

    override fun createViewHolder(view: View, viewType: Int): ViewHolder {
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        val genre = dataSet[position]!!
        if (holder.title != null) {
            holder.title!!.text = genre.name
        }
        if (holder.text != null) {
            holder.text!!.text = MusicUtil.getGenreInfoString(activity, genre)
        }
    }

    override fun loadImage(genre: Genre?, holder: ViewHolder?) {
        if (holder?.image == null) return
        getImageLoader(holder)
            .load(genre!!)
            .into(holder.image!!)
    }

    override fun getName(genre: Genre?): String? {
        return genre?.name
    }

    override fun getSongList(genres: List<Genre?>): List<Song> {
        val songs: MutableList<Song> = ArrayList()
        genres.forEach { genre ->
            songs.addAll(genre?.songs!!) // maybe async in future?
        }
        return songs
    }

    override fun getSectionName(position: Int): String {
        var sectionName: String? = null
        when (genreSortOrder) {
            SortOrder.GenreSortOrder.GENRE_A_Z, SortOrder.GenreSortOrder.GENRE_Z_A -> sectionName =
                dataSet[position]!!.name
        }
        return MusicUtil.getSectionName(sectionName)
    }

    inner class ViewHolder(itemView: View) : MediaEntryViewHolder(itemView) {
        override fun onClick(view: View) {
            if (isInQuickSelectMode) {
                toggleChecked(adapterPosition)
            } else {
                val genre = dataSet[adapterPosition]!!
                mainActivity.navController.toGenreDetail(genre)
            }
        }

        override fun onLongClick(view: View): Boolean {
            toggleChecked(adapterPosition)
            return true
        }

        init {
            if (menu != null) {
                menu!!.visibility = View.GONE
            }
        }
    }
}